From 0d706527d5e7b6d03337a5669e36f160de60df54 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Thu, 14 Nov 2024 10:11:31 +0800
Subject: [PATCH 254/277] fixup! mtk: mt76: mt7996: mt7996 support extra
 option_type

When using single wiphy, the mt7996_hw_phy(hw) only return the first phy.
So the phy->mt76->band_idx cannot indicate the correct band that is being used.
Use mlink->wcid.phy_idx to get the correct band index.

Without this patch, we only use one WED and the token may be insufficient
when the throughput is high.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c | 29 ++++++++++++++---------------
 1 file changed, 14 insertions(+), 15 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 5b04b9e2..7493ec7e 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2505,22 +2505,32 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
 	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
-	struct mt7996_phy *phy = mt7996_hw_phy(hw);
 	struct mtk_wed_device *wed = &dev->mt76.mmio.wed;
 	struct mt7996_bss_conf *mconf;
 	struct mt7996_link_sta *mlink;
 	u8 link_id, dscp = path->mtk_wdma.tid >> 2;
 
+	link_id = msta->pri_link;
+	mconf = rcu_dereference(mvif->link[link_id]);
+	mlink = rcu_dereference(msta->link[link_id]);
+
+	if (!mlink || !mconf || mlink->wcid.idx > MT7996_WTBL_STA)
+		return -EIO;
+
+	if (!test_bit((mvif->qos_map[dscp] + AMPDU_ADDBA_SUCC_SHFT),
+	    &mlink->wcid.ampdu_state))
+		return -EAGAIN;
+
 	if (dev->hif2) {
 		switch (dev->option_type) {
 		case 2:
 			/* mt7996 + 7988d */
-			if (phy->mt76->band_idx == MT_BAND1)
+			if (mlink->wcid.phy_idx == MT_BAND1)
 				wed = &dev->mt76.mmio.wed_hif2;
 			break;
 		default:
-			if ((is_mt7996(&dev->mt76) && phy->mt76->band_idx == MT_BAND2) ||
-			    (is_mt7992(&dev->mt76) && phy->mt76->band_idx == MT_BAND1))
+			if ((is_mt7996(&dev->mt76) && mlink->wcid.phy_idx == MT_BAND2) ||
+			    (is_mt7992(&dev->mt76) && mlink->wcid.phy_idx == MT_BAND1))
 				wed = &dev->mt76.mmio.wed_hif2;
 			break;
 		}
@@ -2529,17 +2539,6 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 	if (!mtk_wed_device_active(wed))
 		return -ENODEV;
 
-	link_id = msta->pri_link;
-	mconf = rcu_dereference(mvif->link[link_id]);
-	mlink = rcu_dereference(msta->link[link_id]);
-
-	if (mlink->wcid.idx > MT7996_WTBL_STA)
-		return -EIO;
-
-	if (!test_bit((mvif->qos_map[dscp] + AMPDU_ADDBA_SUCC_SHFT),
-	    &mlink->wcid.ampdu_state))
-		return -EAGAIN;
-
 	path->type = DEV_PATH_MTK_WDMA;
 	path->dev = ctx->dev;
 	path->mtk_wdma.wdma_idx = wed->wdma_idx;
-- 
2.45.2

